### PR TITLE
Update the adapter for preCICE v3 and add test cases for Elmer-OpenFOAM coupling

### LINKED ISSUE #3: Adapter is not yet ported to preCICE v3
I followed the isntructions of how to compile it. However, I faced an issue with it. Have you seen this error?

`../Coupler_Solver.so: undefined symbol: precicef_is_action_required_

### PR DESCRIPTION
**Pull request by tapegoji**
Description

- Ported Elmer adapter from preCICE v2 to v3.
- Added comprehensive test cases for coupling Elmer with OpenFOAM.
- Ensured compatibility and improved integration with the latest preCICE version.

Changes:

- Updated adapter files to align with preCICE v3 API.
- Included new test cases for Elmer to OpenFOAM coupling. This change required hard-coded naming translation in the Coupler_Solver.F90 code. 
- Enhanced documentation to reflect the changes and new usage instructions.

Testing:

- Successfully ran all new test cases to verify the adapter's functionality with preCICE v3 and OpenFOAM.
- Ensured backward compatibility and verified no breaking changes with existing setups.

Additional Notes:

- Please review the updated documentation and test cases.
- Looking forward to feedback and suggestions for further improvements.

### CODE REVIEW

**Review by uekerman — COMMENTED** [review]
Thanks a lot already! As this change is a bit larger, please be patient with the review :grin: 

Two things already to simplify the review. Please
- Remove the `precice-profiling` folders
- Strip the `tools` to what you actually need

**Comment by tapegoji** [comment]
I have removed the precice-profiling and the extra files in the tools folder. Let me know if you need anything else. I am happy to help make this a great tool. 
Regards,

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]: What was the thought behind this name change? Was it just to make it more clear as to what the variable is?

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
    CHARACTER(LEN=MAX_NAME_LEN)         :: readDataName, writeDataName
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
    meshDim = Mesh % MaxDim
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
        !-----------------Convert to preCICE Naming Conventions
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
        Print *, TRIM(BoundaryName)," ",TRIM(participantName)," ",TRIM(meshName)," ",TRIM(config)
        
         !-----------Identify Vertex on Coupling Interface & Save Coordinates--------------------
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
        CALL Info('CouplerSolver','Initializing preCICE')     
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
        CALL Info('CouplerSolver','Setting up mesh in preCICE')
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]: Why is the variable called `temperature flux_abs` and not just `flux_abs`?

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
        CALL Info('CouplerSolver ', 'Reading data from preCICE')    
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]: Same comment as above.

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion

        !-------------------Copy Write values from Variable to buffer----------------------------
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
        !-------------------Advance preCICE-------------------------------------------------------
```

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]:
```suggestion
        CALL Info('CouplerSolver','preCICE Finalize')
```

File: Flow_Over_Heated_Plate/Fluid-elmer/elmersolver.log
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Fluid-elmer/elmersolver.log:1]: This log file can be removed.

File: Flow_Over_Heated_Plate/README.md
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/README.md:1]: This README would be better suited on main folder level

File: Flow_Over_Heated_Plate/Solid-elmer/elmersolver.log
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Solid-elmer/elmersolver.log:1]: This log file can be removed.

File: Flow_Over_Heated_Plate/precice-config.xml
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/precice-config.xml]:
```suggestion
```

**Review by IshaanDesai — CHANGES_REQUESTED** [review]
@tapegoji thanks a lot for this contribution! Great work already :clap: Please have a look at my review comments. I did not run the test case and also did not have a look at the OpenFOAM things. After we do one iteration of the review, I can try running things. We can then finalize this pull request.

File: Flow_Over_Heated_Plate/Solid-elmer/elmersolver.log
Comment (tapegoji) [reply Flow_Over_Heated_Plate/Solid-elmer/elmersolver.log:1]: Yea I agree

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (tapegoji) [reply Flow_Over_Heated_Plate/Coupler_Solver.F90]: Thad the name of Elmer variable. Not sure with just flux_abs it works. For now leave it to this. I have a new change that is coming and that will change this to magnitude. Elmer has different flux for different physics. By saying temperature flux it means heat flux.

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (tapegoji) [reply Flow_Over_Heated_Plate/Coupler_Solver.F90]: Ok

**Comment by IshaanDesai** [comment]
@tapegoji please let me know when the pull request is ready to be reviewed again. No rush, but I will wait for your ping to have a look again.

**Comment by tapegoji** [comment]
@IshaanDesai, I am a bit confused what I need to do. can you explain what I should do. 
Thank you. I am a bit new and not too familiar with the process.

**Comment by tapegoji** [comment]
I have applied all the changes. let me know if I am missing anything.

**Comment by tapegoji** [comment]
Any update on this?

**Comment by IshaanDesai** [comment]
> Any update on this?

I'll review this week :+1: apologies for the delay, some other urgent things took time.

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90]: Can we remove these lines of code?

File: Flow_Over_Heated_Plate/Fluid-elmer/elmersolver.log
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Fluid-elmer/elmersolver.log:1]: Please delete this file.

File: Flow_Over_Heated_Plate/README.md
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/README.md:1]: Why is this README file duplicated? Unless absolutely required, let us only keep the README at the base adapter level.

File: Flow_Over_Heated_Plate/Solid-elmer/elmersolver.log
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Solid-elmer/elmersolver.log:1]: Please delete this file.

**Review by IshaanDesai — APPROVED** [review]
We are close to merging this. I have added some last comments. For the flow over heated plate tutorial, I see that you have copied over the OpenFOAM files from the [precice/tutorials](https://github.com/precice/tutorials/tree/develop) repository. Instead of copying the files here, a better idea would be to write a small README file here which describes how to copy Elmer files to the tutorials repository and run the tutorial there. This would avoid duplication. If it is not clear how to do this, I or someone else can do this.

Thanks once more for your contribution! Once you address the comments above, we can finalize this.

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (tapegoji) [reply Flow_Over_Heated_Plate/Coupler_Solver.F90]: Yes, Let's delete them. they were in the original one and I did not know why.

File: Flow_Over_Heated_Plate/Fluid-elmer/elmersolver.log
Comment (tapegoji) [reply Flow_Over_Heated_Plate/Fluid-elmer/elmersolver.log:1]: Done!

File: Flow_Over_Heated_Plate/README.md
Comment (tapegoji) [reply Flow_Over_Heated_Plate/README.md:1]: I agree, Ill remove it

File: Flow_Over_Heated_Plate/Solid-elmer/elmersolver.log
Comment (tapegoji) [reply Flow_Over_Heated_Plate/Solid-elmer/elmersolver.log:1]: Done

**Comment by IshaanDesai** [comment]
@tapegoji thanks for making the final changes. Adapter-wise we are ready to merge this. The next step would be to move the files relevant to the flow-over-heated-plate tutorial to the [precice/tutorials](https://github.com/precice/tutorials/tree/develop) repository. I can do this in the next days, and then we can test if everything works. Once we are sure, we can merge this pull request, and also merge Elmer as a new participant to the tutorial.

**Comment by IshaanDesai** [comment]
The Elmer fluid and solid participants for the flow-over-heated-plate tutorial are being added to the appropriate folder in the precice/tutorials repository via: https://github.com/precice/tutorials/pull/565

The cases, and also the OpenFOAM cases copied over for testing have now been deleted. The reference problem has also been removed. The `tools/` folder used for cleanups has also been removed.

I will continue cleaning up this pull request and finalizing the one mentioned above.

File: .gitignore
Comment (IshaanDesai) [inline .gitignore:1]: Lets remove these lines as they are no longer required.

File: Flow_Over_Heated_Plate/Coupler_Solver.F90
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate/Coupler_Solver.F90:1]: I am curious as to why this file is necessary when there is a very similar file `Adapter/Coupler_Solver.F90`. I would suggest moving this file to the `Adapter/` folder, and replacing the older file.

File: Flow_Over_Heated_Plate3D/Reference_Problem/Reference_Problem_Mesh.msh
Comment (IshaanDesai) [inline Flow_Over_Heated_Plate3D/Reference_Problem/Reference_Problem_Mesh.msh:1]: I think we can remove the reference problem after the testing is done. I already removed the reference problem for the 2D case.

**Review by IshaanDesai — COMMENTED** [review]
Here are somethings I noticed while moving the case files to the `precice/tutorials` repository.

**Comment by tapegoji** [comment]
I am good with all the changes. Please go ahead.

**Review by tapegoji — COMMENTED** [review]
sounds good to me

**Comment by IshaanDesai** [comment]
@tapegoji I opened a pull request in the tutorials repository to move the flow over heated plate case files to the respective folder: https://github.com/precice/tutorials/pull/565 I tried running the case there, with both `solid-elmer` and `fluid-elmer`, but the solver did not run as the mesh file was not detected. Can you please have a look and try to run the cases there? I am a bit short on time until the end of next week due to the [upcoming preCICE workshop](https://precice.org/precice-workshop-2024.html).

**Comment by tapegoji** [comment]
did you use the mesh from OpenFoam or did you use the mesh that I provided?

**Comment by IshaanDesai** [comment]
> did you use the mesh from OpenFoam or did you use the mesh that I provided?

I used the mesh which was in the folder of the case in this pull request. Let us continue the discussion in https://github.com/precice/tutorials/pull/565
I am sure it is just something minor. Once the case is running, we can merge everything. Thanks!

**Comment by IshaanDesai** [comment]
Merging this PR and continuing testing https://github.com/precice/tutorials/pull/565 for correctness. The changes in this PR make sense. Minor documentation polishing still needs to be done, but this can also be done on the main branch directly.

**Pull request merged**
